src/librustdoc/passes/html_tags.rs

```diff
@@ -0,0 +1,122 @@
+use super::{span_of_attrs, Pass};
+use crate::clean::*;
+use crate::core::DocContext;
+use crate::fold::DocFolder;
+use crate::html::markdown::opts;
+use rustc_hir::hir_id::HirId;
+use rustc_session::lint;
+use rustc_span::Span;
+use pulldown_cmark::{Event, Parser};
+
+pub const CHECK_UNCLOSED_HTML_TAGS: Pass = Pass {
+    name: "check-unclosed-html-tags",
+    run: check_unclosed_html_tags,
+    description: "detects unclosed HTML tags in doc comments",
+};
+
+struct UnclosedHtmlTagsLinter<'a, 'tcx> {
+    cx: &'a DocContext<'tcx>,
+}
+
+impl<'a, 'tcx> UnclosedHtmlTagsLinter<'a, 'tcx> {
+    fn new(cx: &'a DocContext<'tcx>) -> Self {
+        UnclosedHtmlTagsLinter { cx }
+    }
+}
+
+pub fn check_unclosed_html_tags(krate: Crate, cx: &DocContext<'_>) -> Crate {
+    let mut coll = UnclosedHtmlTagsLinter::new(cx);
+
+    coll.fold_crate(krate)
+}
+
+const ALLOWED_UNCLOSED: &[&str] = &[
+    "area", "base", "br", "col", "embed", "hr", "img", "input", "keygen", "link", "meta", "param",
+    "source", "track", "wbr",
+];
+
+fn drop_tag(cx: &DocContext<'_>, tags: &mut Vec<String>, tag_name: String, hir_id: HirId, sp: Span) {
+    // In case the tag wasn't in the queue, I guess it's not really an issue...
+    if let Some(pos) = tags.iter().position(|t| *t == tag_name) {
+        for _ in pos + 1..tags.len() {
+            if ALLOWED_UNCLOSED.iter().find(|&at| at == &tags[pos + 1]).is_some() {
+                continue;
+            }
+            cx.tcx.struct_span_lint_hir(
+                lint::builtin::UNCLOSED_HTML_TAGS,
+                hir_id,
+                sp,
+                |lint| lint.build(&format!("Unclosed HTML tag `{}`", tags[pos])).emit(),
+            );
+            tags.remove(pos + 1);
+        }
+        tags.remove(pos);
+    }
+}
+
+fn extract_tag(cx: &DocContext<'_>, tags: &mut Vec<String>, text: &str, hir_id: HirId, sp: Span) {
+    let mut iter = text.chars().peekable();
+
+    while let Some(c) = iter.next() {
+        if c == '<' {
+            let mut tag_name = String::new();
+            let mut is_closing = false;
+            while let Some(&c) = iter.peek() {
+                if c == '/' && tag_name.is_empty() {
+                    is_closing = true;
+                } else if c.is_ascii_alphanumeric() && !c.is_ascii_uppercase() {
+                    tag_name.push(c);
```

Comment:
```suggestion
                } else if c.is_ascii_alphanumeric() && !c.is_ascii_uppercase() {
                    // <script>
                    tag_name.push(c);
```

Reply:
No, it can also be `</script>`.

Reply:
Oh I see this runs for _every_ character in the loop, so it runs for both `<script>` and `</script>`. Got it.